src/hb-ot-metrics.cc

```diff
@@ -0,0 +1,117 @@
+ * ON AN "AS IS" BASIS, AND THE COPYRIGHT HOLDER HAS NO OBLIGATION TO
+ * PROVIDE MAINTENANCE, SUPPORT, UPDATES, ENHANCEMENTS, OR MODIFICATIONS.
+ */
+
+#include "hb.hh"
+
+#include "hb-ot-gasp-table.hh"
+#include "hb-ot-os2-table.hh"
+#include "hb-ot-post-table.hh"
+#include "hb-ot-hmtx-table.hh"
+#include "hb-ot-var-mvar-table.hh"
+
+#include "hb-ot-face.hh"
+
+#if 0
+static bool
+_get_gasp (hb_font_t *font, unsigned int i, hb_position_t *result,
+	   hb_ot_metrics_t metrics_tag)
+{
+  const OT::GaspRange& range = font->face->table.gasp->get_gasp_range (i);
+  if (&range == &Null (OT::GaspRange)) return false;
+  if (result) *result = font->em_scale_y (range.rangeMaxPPEM +
+					  font->face->table.MVAR->get_var (metrics_tag, nullptr, 0));
+  return true;
+}
+#endif
+
+/**
+ * hb_ot_metrics_get:
+ * @font:
+ * @metrics_tag:
+ * @position:    (out) (optional):
+ *
+ * Returns: Whether found the requested metrics
+ *
+ * Since: REPLACEME
+ **/
+hb_bool_t
+hb_ot_metrics_get (hb_font_t       *font,
+		   hb_ot_metrics_t  metrics_tag,
+		   hb_position_t   *position     /* OUT.  May be NULL. */)
+{
+  switch (metrics_tag)
+  {
+#define GET_METRIC(TABLE, ATTR, DIR) \
+  (font->face->table.TABLE->has_data () && \
+    (position && (*position = font->em_scale_##DIR (font->face->table.TABLE->ATTR + \
+						    font->face->table.MVAR->get_var (metrics_tag, nullptr, 0))), true))
+#define GET_X_METRIC(TABLE, ATTR) GET_METRIC (TABLE, ATTR, x)
+#define GET_Y_METRIC(TABLE, ATTR) GET_METRIC (TABLE, ATTR, y)
+
+  case HB_OT_METRICS_HORIZONTAL_ASCENDER:         return GET_Y_METRIC (hmtx, ascender);
+  case HB_OT_METRICS_HORIZONTAL_DESCENDER:        return GET_Y_METRIC (hmtx, descender);
+  case HB_OT_METRICS_HORIZONTAL_LINE_GAP:         return GET_Y_METRIC (hmtx, line_gap);
+  case HB_OT_METRICS_HORIZONTAL_CLIPPING_ASCENT:  return GET_Y_METRIC (OS2, usWinAscent);
+  case HB_OT_METRICS_HORIZONTAL_CLIPPING_DESCENT: return GET_Y_METRIC (OS2, usWinDescent);
+  case HB_OT_METRICS_VERTICAL_ASCENDER:           return GET_X_METRIC (vmtx, ascender);
+  case HB_OT_METRICS_VERTICAL_DESCENDER:          return GET_X_METRIC (vmtx, descender);
+  case HB_OT_METRICS_VERTICAL_LINE_GAP:           return GET_X_METRIC (vmtx, line_gap);
+  case HB_OT_METRICS_HORIZONTAL_CARET_RISE:       return GET_Y_METRIC (hhea, caretSlopeRise);
+  case HB_OT_METRICS_HORIZONTAL_CARET_RUN:        return GET_X_METRIC (hhea, caretSlopeRun);
+  case HB_OT_METRICS_HORIZONTAL_CARET_OFFSET:     return GET_X_METRIC (hhea, caretOffset);
+  case HB_OT_METRICS_VERTICAL_CARET_RISE:         return GET_X_METRIC (vhea, caretSlopeRise);
+  case HB_OT_METRICS_VERTICAL_CARET_RUN:          return GET_Y_METRIC (vhea, caretSlopeRun);
+  case HB_OT_METRICS_VERTICAL_CARET_OFFSET:       return GET_Y_METRIC (vhea, caretOffset);
+  case HB_OT_METRICS_X_HEIGHT:                    return GET_Y_METRIC (OS2->v2 (), sxHeight);
+  case HB_OT_METRICS_CAP_HEIGHT:                  return GET_Y_METRIC (OS2->v2 (), sCapHeight);
```

Comment:
I'm confused.  So, `->v2 ()` works now, even though that's a reference and you expect a pointer here?

Reply:
Oh!  I wasn't suggesting you overload it.  Which, honestly, doesn't make sense.  I was literally meaning if you write `v2 ().operator ->`.  Don't know if that works even if you haven't overloaded it.  Fine for now I guess.  I just don't like a particular use of a class dictating the design of the class.